src/lib.rs

```diff
@@ -837,6 +837,87 @@ pub enum Format {
     NoSpaces,
 }
 
+/// The representation of a color stop for a `ColorScale`.
+/// The position defines where the color is placed from left (0.0) to right (1.0).
+#[derive(Debug, Clone)]
+struct ColorStop {
+    color: Color,
+    position: Fraction,
+}
+
+/// The representation of a color scale.
+/// The first `ColorStop` (position 0.0) defines the left end color.
+/// The last `ColorStop` (position 1.0) defines the right end color.
+#[derive(Debug, Clone)]
+pub struct ColorScale {
+    color_stops: Vec<ColorStop>,
+}
+
+impl ColorScale {
+    /// Create an empty `ColorScale`.
+    pub fn empty() -> Self {
+        Self {
+            color_stops: Vec::new(),
+        }
+    }
+
+    /// Add a `Color` at the given position.
+    pub fn add_stop(&mut self, color: Color, position: Fraction) -> &mut Self {
+        let color_stop = ColorStop { color, position };
+
+        let next_index = self
+            .color_stops
+            .iter()
+            .position(|c| position.value() <= c.position.value());
+
+        let index = match next_index {
+            Some(0) => 0,
+            Some(next) => next - 1,
+            None => self.color_stops.len(),
+        };
```

Comment:
I think the logic is wrong here. `next - 1` should be `next`. Imagine what happens if you first insert `0.0`, then `1.0`, then `0.5`:
* the `0.0` would be inserted at index `0` (the `None` case) => [0.0]
* the `1.0` would be inserted at index `0` (the `Some(0)` case) => [0.0, 1.0]
* the `0.5` would be inserted at index `0` (the `Some(next)` case with `next == 1`. This would result in [0.5, 0.0, 1.0].

The code can be fixed by using:
``` rust
let index = next_index.unwrap_or(self.color_stops.len());
```

Reply:
It would be great if there would be a test for this. The `color_scale_sample_position` currently fails if you change the order of `.add_stop` calls.

Reply:
I've changed the `add_stop` order on some tests to catch this issue.